File: day10/part2.py
# Read file input
with open("input.txt", "r") as f:
    data = [i.rstrip('\n') for i in f]

points = {')':1,']':2,'}':3,'>':4}
closing = {'(':')','[':']','{':'}','<':'>'}

# Calculate completion score for each incomplete line
completion_scores = []
for line in data:
    current = [] # Stack containing necessary closing characters
    for c in line:
        if c in closing:
            current.append(closing[c])
        else:
            if current and current[-1] == c:
                current.pop()
            else:
                current = []
                break
    # Incomplete line
    if current:
        score = 0
        for c in current[::-1]:
            score *= 5
            score += points[c]
        completion_scores.append(score)

# Take median score
completion_scores.sort()
middle_score = completion_scores[len(completion_scores)//2]

print(f"Result: {middle_score}")

File: day10/part2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <unordered_map>
#include <algorithm>
using namespace std;
#define ll int64_t

int main() {
	freopen("input.txt","r",stdin);

	// Read file input
	vector<string> data;
	string s;
	while (cin >> s)
		data.push_back(s);

	unordered_map<char,int> points = {
		{')', 1},
		{']', 2},
		{'}', 3},
		{'>', 4}
	};
	unordered_map<char,char> closing = {
		{'(', ')'},
		{'[', ']'},
		{'{', '}'},
		{'<', '>'}
	};

	// Calculate score for each line
	vector<ll> completion_scores;
	for (string line : data) {
		stack<int> current;
		for (char c : line) {
			if (closing.find(c) != closing.end()) {
				current.push(closing[c]);
			} else {
				if (!current.empty() && current.top() == c) {
					current.pop();
				} else {
					current = {};
					break;
				}
			}
		}
		// Calculate completion score (0 for complete lines)
		ll score = 0;
		while (!current.empty()) {
			score *= 5;
			score += points[current.top()];
			current.pop();
		}
		if (score)
			completion_scores.push_back(score);
	}
	// Take median score
	sort(completion_scores.begin(), completion_scores.end());
	ll middle_score = completion_scores[completion_scores.size()/2];

	cout << "Result: " << middle_score << '\n';
}
